### ISSUE TITLE
how to add enum in templates , its not showing combo box for templates

### ISSUE THREAD

**Issue by sumancholleti** [josdejong/jsoneditor#473]

**Comment by josdejong** [comment]
These enum dropdowns are a feature that you can use by providing a JSON Schema enumeration, so you could use it in combination with a template.

**Comment by sumancholleti** [comment]
var schema = {
	    "title": "Example Composite Schema",
	    "oneOf": [
	    	{
	      	"type": "object",
	      	"properties": {
	        	"enumerable_prop": {
	          		"type": "string",
	              "enum": ["first", "second", "third"]
	          }
	        }
	      },
	      {
	      	"type": "array"
	      }
	    ]
	  };
	  
var options = {
name: 'UconnectMessage',
schema: schema,
templates: [
    {
        text: 'text',
        title: 'Insert a Message',
        className: 'any',
        field: 'notify,
        value: {
        	enumerable_prop:null
        }
    }



Can you give an example, I tried this , its not showing dropdown

**Comment by josdejong** [comment]
Here you go:

http://jsbin.com/ciqifub/edit?html,output

**Comment by sumancholleti** [comment]
Can json be an object instead?

**Comment by josdejong** [comment]
yes. just play around with it and you will see

**Comment by sumancholleti** [comment]
I tried and it didn't work

**Comment by josdejong** [comment]
Keep trying :)

You will have to adjust both JSON and JSON schema, these need to be aligned with each other.

**Comment by sumancholleti** [comment]
No luck yet
var schema = {
        "type": "object",
        "properties": {
          "gender": {
            "enum": ["male", "female"]
        }
      }
    };


    var json = {};
      
    
    var templates = [
      {
        text: 'Employee',
        title: 'Insert a new employee',
        className: 'uconnect-any',
        field: 'employee',
        value: {
          gender: '',
        }
      }
    ]
    
    var options = {
      schema: schema,
      templates: templates
    };

    // create the editor
    var container = document.getElementById('jsoneditor');
    var editor = new JSONEditor(container, options, json);

**Comment by josdejong** [comment]
I think the difficulty here is that the enum dropdown only works when the JSON matches the JSONSchema, whilst a template can be inserted everywhere. In your example code the enum for `gender` only works for a property in the root object, and not for nested objects. When inserting an employee via the template, you create a nested object - which is not matching the JSON Schema enum.

**Comment by sumancholleti** [comment]
I didn't get you

**Comment by josdejong** [comment]
The schema that you post:
```
var schema = {
  "type": "object",
    "properties": {
    "gender": {
      "enum": ["male", "female"]
    }
  }
};
```
will only create a dropdown for the field `gender` in the root of the JSON object like:
```
{
  "gender": "male"
}
```
It will NOT create a dropdown not for nested objects like here:
```
{
  "someNestedObject": {
    "gender": "male"
  }
}
```
if you need that, you will need to adjust your JSON Schema to have the definition of `gender` there where it will be created in your JSON object. If you want a dropdown for `gender` everywhere in the object (on all levels), you will have to define a JSON Schema with recursion like discussed [here](https://stackoverflow.com/questions/20752716/json-schema-recursive-schema-definition).

**Comment by ortrud** [comment]
I have followed this conversation and tried everything you suggested, but I can not get a dropdown to show up anywhere but on the top level of the schema for an enum property like gender.
``let demographicsSchema = {
	"type": "object",
       	"properties": {
       		"firstName": { "type": "string", "minLength" : 1 },
          	//"lastName": { "type" : ["string","null"] },
       		//"gender": { "enum": ["Male", "Female"] },
          	"lastName": { "type" : "string" },
		"gender" : {"$ref" : "gender"  } ,
       		"age": { "type": "number", "minimum" : 0, "default" : 40},
       		"maritalStatus": { "type": "string", "enum" : ["M","S","DV","WD","MA","NM","SE"] },
       		"cin": { "type": "string" },
       		"sufficientInfoForCIN": { "type": "string" },
       		"middleNameOrInitial": { "type": "string" },
       		"nameSuffix": { "type": "string" },
       		"nameVerified": { "type": "string" },
       		"ssn": { "type": "string", minLength:9, maxLength:9 },
       		"ssnStatus": { "type": "string", "default" : "VR" },
	},
	"required": ["firstName", "gender", "age"]
};

let genderSchema = {
	"type": "string",
       	"enum": ["Male", "Female"] 
};

Do you have plans to make dropdown available on any level?   
Also, what is the relation between jsoneditor and json-editor projects.  I like jsoneditir, but json-editor does handle enums a sdropdowns on any level of the schema

**Comment by josdejong** [comment]
@ortrud have you tried the example that I posted [here](https://github.com/josdejong/jsoneditor/issues/473#issuecomment-335273889)? http://jsbin.com/ciqifub/edit?html,output

There is no relation between jsoneditor and json-editor except that they are both editors for JSON :)

If your json schema defines enums on any level, then JSONEditor will display them on any level - that's a matter of how you define your json schema.

**Comment by SteveNeithardt** [comment]
![screenshot_2018-08-22 js bin](https://user-images.githubusercontent.com/22314807/44456520-0b538080-a601-11e8-84ab-25cd1072ca9b.png)

I have the same issue, where the enums drop downs never seem to appear within objects in arrays.

Here's the screenshot of the example you posted on jsbin.

**Comment by josdejong** [comment]
Hm, that's odd. Will look into it, thanks for reporting @SteveNeithardt.

**Comment by chriscarreau** [comment]
Hi, I'm still running into this issue (or a similar one) in the current version (5.32.4) but only in the specific case where my schema has an object with additionalProperties having an enum.

Here's an example: https://jsbin.com/vuyabuyiji/edit?html,output

![image](https://user-images.githubusercontent.com/6654898/55996989-29045180-5c87-11e9-8671-78095fe6f1bb.png)
![image](https://user-images.githubusercontent.com/6654898/55997006-391c3100-5c87-11e9-84cd-00c8c788e7b4.png)

For the error to show up you have to add additional properties under the "testObj" object.

As you can see in this example, the validation works, telling me I have to choose one of the enum values,  but the combobox/select does not appear for the addtional properties.

Thank you for looking into it 😄

**Comment by josdejong** [comment]
@chriscarreau thanks for reporting (with a reproducible jsbin). Not sure, but I *think* the enum dropdowns don't (yet) reckon with `additionalProperties`.

**Comment by josdejong** [comment]
Someone interested in looking into rendering an `enum` dropdown for `additionalProperties`?

**Comment by yotamgod** [comment]
I was wondering if dropdowns for schemas with additionalProperties have been addressed yet?
I don't get a dropdown for enum properties inside an additionalProperties in version `9.1.2`.
My schema is as follows:
```
{
  "type": "object",
  "title": "",
  "description": "",
  "default": {},
  "additionalProperties": {
    "type": "object",
    "properties": {
      "day": {
        "type": "string",
        "enum": [
          "sunday",
          "monday",
          "tuesday"
        ]
      }
    }
  }
}
```

**Comment by josdejong** [comment]
@yotamgod I'm not sure. I've just published `v9.1.3` which contains a fix by @maufl which _may_ be related, see #1158 and #1161. Can you try out `jsoneditor@9.1.3` to see whether your issue was accidentally fixed as a side effect?

**Comment by yotamgod** [comment]
Sadly it hasn't :(

**Comment by josdejong** [comment]
Thanks for checking 👍 . So we still have to do some debugging here.

**Comment by josdejong** [comment]
This should be fixed now in `v9.5.3`, see #1355. 

If not fixed, please re-open this issue again.
